File: backend/shared/models/subscription.ts
import { query } from '../utils/database';

/**
 * Subscription interface
 */
export interface Subscription {
  id: string;
  subscriber_id: string;
  schema_id: string;
  webhook_url: string;
  webhook_secret: string;
  auth_type: string;
  auth_config?: any;
  custom_headers?: any;
  enabled: boolean;
  max_retries: number;
  backoff_strategy: 'exponential' | 'linear' | 'constant';
  initial_delay_ms: number;
  timeout_ms: number;
  payload_filter?: any;
  status: 'active' | 'paused' | 'suspended' | 'cancelled';
  total_deliveries: number;
  successful_deliveries: number;
  failed_deliveries: number;
  avg_latency_ms: number;
  consecutive_failures: number;
  last_delivery_at?: Date;
  last_success_at?: Date;
  last_failure_at?: Date;
  circuit_breaker_open_until?: Date;
  approval_status: 'pending' | 'approved' | 'rejected';
  approved_by?: string;
  approved_at?: Date;
  created_at: Date;
  updated_at: Date;
  cancelled_at?: Date;
}

/**
 * Create subscription DTO
 */
export interface CreateSubscriptionDTO {
  subscriber_id: string;
  schema_id: string;
  webhook_url: string;
  webhook_secret: string;
  auth_type?: string;
  auth_config?: any;
  custom_headers?: any;
  enabled?: boolean;
  max_retries?: number;
  backoff_strategy?: 'exponential' | 'linear' | 'constant';
  initial_delay_ms?: number;
  timeout_ms?: number;
  payload_filter?: any;
}

/**
 * Update subscription DTO
 */
export interface UpdateSubscriptionDTO {
  webhook_url?: string;
  auth_type?: string;
  auth_config?: any;
  custom_headers?: any;
  enabled?: boolean;
  max_retries?: number;
  backoff_strategy?: 'exponential' | 'linear' | 'constant';
  initial_delay_ms?: number;
  timeout_ms?: number;
  payload_filter?: any;
  status?: 'active' | 'paused' | 'suspended';
}

/**
 * Create a new subscription
 */
export async function createSubscription(data: CreateSubscriptionDTO): Promise<Subscription> {
  const result = await query<Subscription>(
    `INSERT INTO subscriptions (
      subscriber_id, schema_id, webhook_url, webhook_secret, auth_type, auth_config,
      custom_headers, enabled, max_retries, backoff_strategy, initial_delay_ms,
      timeout_ms, payload_filter
    ) VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12, $13)
    RETURNING *`,
    [
      data.subscriber_id,
      data.schema_id,
      data.webhook_url,
      data.webhook_secret,
      data.auth_type || 'hmac',
      data.auth_config ? JSON.stringify(data.auth_config) : null,
      data.custom_headers ? JSON.stringify(data.custom_headers) : null,
      data.enabled !== false,
      data.max_retries ?? 3,
      data.backoff_strategy || 'exponential',
      data.initial_delay_ms ?? 1000,
      data.timeout_ms ?? 30000,
      data.payload_filter ? JSON.stringify(data.payload_filter) : null,
    ]
  );

  return result.rows[0];
}

/**
 * Get subscription by ID
 */
export async function getSubscriptionById(id: string): Promise<Subscription | null> {
  const result = await query<Subscription>(
    'SELECT * FROM subscriptions WHERE id = $1 AND cancelled_at IS NULL',
    [id]
  );

  return result.rows[0] || null;
}

/**
 * Get subscription by subscriber and schema
 */
export async function getSubscriptionBySubscriberAndSchema(
  subscriberId: string,
  schemaId: string
): Promise<Subscription | null> {
  const result = await query<Subscription>(
    'SELECT * FROM subscriptions WHERE subscriber_id = $1 AND schema_id = $2 AND cancelled_at IS NULL',
    [subscriberId, schemaId]
  );

  return result.rows[0] || null;
}

/**
 * List subscriptions for a subscriber
 */
export async function listSubscriptionsForSubscriber(
  subscriberId: string,
  filters: {
    status?: string;
    enabled?: boolean;
  } = {},
  page: number = 1,
  limit: number = 20
): Promise<{ subscriptions: Subscription[]; total: number }> {
  const offset = (page - 1) * limit;
  const conditions: string[] = ['subscriber_id = $1', 'cancelled_at IS NULL'];
  const params: any[] = [subscriberId];
  let paramIndex = 2;

  if (filters.status) {
    conditions.push(`status = $${paramIndex}`);
    params.push(filters.status);
    paramIndex++;
  }

  if (filters.enabled !== undefined) {
    conditions.push(`enabled = $${paramIndex}`);
    params.push(filters.enabled);
    paramIndex++;
  }

  const whereClause = conditions.join(' AND ');

  // Get total count
  const countResult = await query(
    `SELECT COUNT(*) as total FROM subscriptions WHERE ${whereClause}`,
    params
  );
  const total = parseInt(countResult.rows[0].total, 10);

  // Get paginated results
  const subscriptionsResult = await query<Subscription>(
    `SELECT * FROM subscriptions
     WHERE ${whereClause}
     ORDER BY created_at DESC
     LIMIT $${paramIndex} OFFSET $${paramIndex + 1}`,
    [...params, limit, offset]
  );

  return {
    subscriptions: subscriptionsResult.rows,
    total,
  };
}

/**
 * List active subscriptions for a schema (for event delivery)
 */
export async function listActiveSubscriptionsForSchema(schemaId: string): Promise<Subscription[]> {
  const result = await query<Subscription>(
    `SELECT * FROM subscriptions
     WHERE schema_id = $1
       AND enabled = true
       AND status = 'active'
       AND cancelled_at IS NULL
       AND (circuit_breaker_open_until IS NULL OR circuit_breaker_open_until < CURRENT_TIMESTAMP)`,
    [schemaId]
  );

  return result.rows;
}

/**
 * Get subscription with schema and subscriber details
 */
export async function getSubscriptionWithDetails(subscriptionId: string): Promise<{
  subscription: Subscription;
  schema: { id: string; name: string; event_type: string; version: string };
  subscriber: { id: string; name: string; email: string };
} | null> {
  const result = await query(
    `SELECT
      s.*,
      sch.id as schema_id,
      sch.name as schema_name,
      sch.event_type as schema_event_type,
      sch.version as schema_version,
      sub.id as subscriber_id,
      sub.name as subscriber_name,
      sub.email as subscriber_email
     FROM subscriptions s
     JOIN schemas sch ON sch.id = s.schema_id
     JOIN subscribers sub ON sub.id = s.subscriber_id
     WHERE s.id = $1 AND s.cancelled_at IS NULL`,
    [subscriptionId]
  );

  if (!result.rows[0]) {
    return null;
  }

  const row = result.rows[0];

  return {
    subscription: {
      id: row.id,
      subscriber_id: row.subscriber_id,
      schema_id: row.schema_id,
      webhook_url: row.webhook_url,
      webhook_secret: row.webhook_secret,
      auth_type: row.auth_type,
      auth_config: row.auth_config,
      custom_headers: row.custom_headers,
      enabled: row.enabled,
      max_retries: row.max_retries,
      backoff_strategy: row.backoff_strategy,
      initial_delay_ms: row.initial_delay_ms,
      timeout_ms: row.timeout_ms,
      payload_filter: row.payload_filter,
      status: row.status,
      total_deliveries: row.total_deliveries,
      successful_deliveries: row.successful_deliveries,
      failed_deliveries: row.failed_deliveries,
      avg_latency_ms: row.avg_latency_ms,
      consecutive_failures: row.consecutive_failures,
      last_delivery_at: row.last_delivery_at,
      last_success_at: row.last_success_at,
      last_failure_at: row.last_failure_at,
      circuit_breaker_open_until: row.circuit_breaker_open_until,
      approval_status: row.approval_status,
      approved_by: row.approved_by,
      approved_at: row.approved_at,
      created_at: row.created_at,
      updated_at: row.updated_at,
      cancelled_at: row.cancelled_at,
    },
    schema: {
      id: row.schema_id,
      name: row.schema_name,
      event_type: row.schema_event_type,
      version: row.schema_version,
    },
    subscriber: {
      id: row.subscriber_id,
      name: row.subscriber_name,
      email: row.subscriber_email,
    },
  };
}

/**
 * Update subscription
 */
export async function updateSubscription(
  id: string,
  data: UpdateSubscriptionDTO
): Promise<Subscription | null> {
  const updates: string[] = [];
  const params: any[] = [];
  let paramIndex = 1;

  if (data.webhook_url !== undefined) {
    updates.push(`webhook_url = $${paramIndex}`);
    params.push(data.webhook_url);
    paramIndex++;
  }

  if (data.auth_type !== undefined) {
    updates.push(`auth_type = $${paramIndex}`);
    params.push(data.auth_type);
    paramIndex++;
  }

  if (data.auth_config !== undefined) {
    updates.push(`auth_config = $${paramIndex}`);
    params.push(JSON.stringify(data.auth_config));
    paramIndex++;
  }

  if (data.custom_headers !== undefined) {
    updates.push(`custom_headers = $${paramIndex}`);
    params.push(JSON.stringify(data.custom_headers));
    paramIndex++;
  }

  if (data.enabled !== undefined) {
    updates.push(`enabled = $${paramIndex}`);
    params.push(data.enabled);
    paramIndex++;
  }

  if (data.max_retries !== undefined) {
    updates.push(`max_retries = $${paramIndex}`);
    params.push(data.max_retries);
    paramIndex++;
  }

  if (data.backoff_strategy !== undefined) {
    updates.push(`backoff_strategy = $${paramIndex}`);
    params.push(data.backoff_strategy);
    paramIndex++;
  }

  if (data.initial_delay_ms !== undefined) {
    updates.push(`initial_delay_ms = $${paramIndex}`);
    params.push(data.initial_delay_ms);
    paramIndex++;
  }

  if (data.timeout_ms !== undefined) {
    updates.push(`timeout_ms = $${paramIndex}`);
    params.push(data.timeout_ms);
    paramIndex++;
  }

  if (data.payload_filter !== undefined) {
    updates.push(`payload_filter = $${paramIndex}`);
    params.push(JSON.stringify(data.payload_filter));
    paramIndex++;
  }

  if (data.status !== undefined) {
    updates.push(`status = $${paramIndex}`);
    params.push(data.status);
    paramIndex++;
  }

  if (updates.length === 0) {
    return null;
  }

  updates.push(`updated_at = CURRENT_TIMESTAMP`);

  const result = await query<Subscription>(
    `UPDATE subscriptions
     SET ${updates.join(', ')}
     WHERE id = $${paramIndex} AND cancelled_at IS NULL
     RETURNING *`,
    [...params, id]
  );

  return result.rows[0] || null;
}

/**
 * Cancel subscription
 */
export async function cancelSubscription(id: string): Promise<boolean> {
  const result = await query(
    `UPDATE subscriptions
     SET status = 'cancelled', cancelled_at = CURRENT_TIMESTAMP, updated_at = CURRENT_TIMESTAMP
     WHERE id = $1 AND cancelled_at IS NULL`,
    [id]
  );

  return result.rowCount > 0;
}

/**
 * Approve subscription
 */
export async function approveSubscription(
  id: string,
  approvedBy: string
): Promise<Subscription | null> {
  const result = await query<Subscription>(
    `UPDATE subscriptions
     SET approval_status = 'approved',
         approved_by = $1,
         approved_at = CURRENT_TIMESTAMP,
         updated_at = CURRENT_TIMESTAMP
     WHERE id = $2
     RETURNING *`,
    [approvedBy, id]
  );

  return result.rows[0] || null;
}

/**
 * Reject subscription
 */
export async function rejectSubscription(
  id: string,
  reviewedBy: string
): Promise<Subscription | null> {
  const result = await query<Subscription>(
    `UPDATE subscriptions
     SET approval_status = 'rejected',
         approved_by = $1,
         approved_at = CURRENT_TIMESTAMP,
         status = 'suspended',
         updated_at = CURRENT_TIMESTAMP
     WHERE id = $2
     RETURNING *`,
    [reviewedBy, id]
  );

  return result.rows[0] || null;
}

/**
 * Open circuit breaker (pause deliveries temporarily)
 */
export async function openCircuitBreaker(
  subscriptionId: string,
  durationMinutes: number = 30
): Promise<void> {
  await query(
    `UPDATE subscriptions
     SET circuit_breaker_open_until = CURRENT_TIMESTAMP + INTERVAL '${durationMinutes} minutes'
     WHERE id = $1`,
    [subscriptionId]
  );
}

File: backendpy/shared/models/subscription.py
from ..utils.database import query, query_one


def list_subscriptions(filters=None, page=1, limit=20):
    """
    List subscriptions with pagination and schema details

    Args:
        filters: Optional dict with subscriber_id, status, enabled
        page: Page number (1-indexed)
        limit: Number of results per page

    Returns:
        Dict with 'subscriptions' array and 'total' count
    """
    offset = (page - 1) * limit
    conditions = []
    params = []

    # Build WHERE conditions
    if filters:
        if filters.get('subscriber_id'):
            conditions.append("s.subscriber_id = %s")
            params.append(filters['subscriber_id'])

        if filters.get('status'):
            conditions.append("s.status = %s")
            params.append(filters['status'])

        if filters.get('enabled') is not None:
            conditions.append("s.enabled = %s")
            params.append(filters['enabled'])

    # Always filter out cancelled subscriptions
    conditions.append("s.cancelled_at IS NULL")

    where_clause = f"WHERE {' AND '.join(conditions)}" if conditions else ""

    # Get total count
    count_sql = f"SELECT COUNT(*) as total FROM subscriptions s {where_clause}"
    total_result = query_one(count_sql, params)
    total = total_result['total'] if total_result else 0

    # Get paginated results with schema details
    subscriptions_sql = f"""
        SELECT
            s.id,
            s.subscriber_id,
            s.schema_id,
            s.webhook_url,
            s.max_retries,
            s.backoff_strategy,
            s.enabled,
            s.status,
            s.total_deliveries,
            s.successful_deliveries,
            s.failed_deliveries,
            s.created_at,
            sch.id as schema_pk_id,
            sch.name as schema_name,
            sch.event_type as schema_event_type,
            sch.version as schema_version
        FROM subscriptions s
        LEFT JOIN schemas sch ON sch.id = s.schema_id
        {where_clause}
        ORDER BY s.created_at DESC
        LIMIT %s OFFSET %s
    """
    params.extend([limit, offset])
    subscriptions = query(subscriptions_sql, params)

    return {
        'subscriptions': [dict(sub) for sub in subscriptions] if subscriptions else [],
        'total': total
    }


def get_subscription_by_id(subscription_id):
    """Get subscription by ID"""
    sql = """
        SELECT
            s.*,
            sch.id as schema_pk_id,
            sch.name as schema_name,
            sch.event_type as schema_event_type,
            sch.version as schema_version
        FROM subscriptions s
        LEFT JOIN schemas sch ON sch.id = s.schema_id
        WHERE s.id = %s AND s.cancelled_at IS NULL
    """
    result = query_one(sql, (subscription_id,))
    return dict(result) if result else None


def get_subscriber_by_api_key(api_key):
    """Get subscriber by API key (using hash comparison)"""
    # For now, using simple comparison - should use hash in production
    sql = """
        SELECT * FROM subscribers
        WHERE api_key = %s AND status = 'active'
    """
    result = query_one(sql, (api_key,))
    return dict(result) if result else None


def create_subscription(data):
    """
    Create a new subscription

    Args:
        data: Dict with subscription fields (subscriber_id, schema_id, webhook_url,
              webhook_secret, max_retries, backoff_strategy, enabled)

    Returns:
        Dict with created subscription
    """
    sql = """
        INSERT INTO subscriptions (
            subscriber_id, schema_id, webhook_url, webhook_secret,
            max_retries, backoff_strategy, enabled, auth_type, status
        ) VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s)
        RETURNING *
    """

    params = (
        data['subscriber_id'],
        data['schema_id'],
        data['webhook_url'],
        data.get('webhook_secret', ''),
        data.get('max_retries', 3),
        data.get('backoff_strategy', 'exponential'),
        data.get('enabled', True),
        data.get('auth_type', 'hmac'),
        data.get('status', 'active')
    )

    result = query_one(sql, params)
    return dict(result) if result else None


def list_subscribers(filters=None, page=1, limit=20):
    """
    List subscribers with pagination

    Args:
        filters: Optional dict with status, search
        page: Page number (1-indexed)
        limit: Number of results per page

    Returns:
        Dict with 'subscribers' array and 'total' count
    """
    offset = (page - 1) * limit
    conditions = []
    params = []

    if filters:
        if filters.get('status'):
            conditions.append("status = %s")
            params.append(filters['status'])

        if filters.get('search'):
            conditions.append("(name ILIKE %s OR email ILIKE %s)")
            search_term = f"%{filters['search']}%"
            params.extend([search_term, search_term])

    where_clause = f"WHERE {' AND '.join(conditions)}" if conditions else ""

    # Get total count
    count_sql = f"SELECT COUNT(*) as total FROM subscribers {where_clause}"
    total_result = query_one(count_sql, params)
    total = total_result['total'] if total_result else 0

    # Get paginated results
    subscribers_sql = f"""
        SELECT id, name, email, webhook_url, webhook_secret, status, created_at, updated_at
        FROM subscribers
        {where_clause}
        ORDER BY created_at DESC
        LIMIT %s OFFSET %s
    """
    params.extend([limit, offset])
    subscribers = query(subscribers_sql, params)

    return {
        'subscribers': [dict(sub) for sub in subscribers] if subscribers else [],
        'total': total
    }


def update_subscriber(subscriber_id, data):
    """
    Update a subscriber

    Args:
        subscriber_id: Subscriber ID
        data: Dict with fields to update (name, email, webhookUrl, status)

    Returns:
        Dict with updated subscriber or None if not found
    """
    # Build update query dynamically based on provided fields
    update_fields = []
    params = []

    if 'name' in data:
        update_fields.append('name = %s')
        params.append(data['name'])

    if 'email' in data:
        update_fields.append('email = %s')
        params.append(data['email'])

    if 'webhookUrl' in data or 'webhook_url' in data:
        update_fields.append('webhook_url = %s')
        params.append(data.get('webhookUrl') or data.get('webhook_url'))

    if 'status' in data:
        update_fields.append('status = %s')
        params.append(data['status'])

    if not update_fields:
        # No fields to update, just return current subscriber
        sql = "SELECT * FROM subscribers WHERE id = %s"
        result = query_one(sql, (subscriber_id,))
        return dict(result) if result else None

    # Add updated_at
    update_fields.append('updated_at = CURRENT_TIMESTAMP')

    # Add subscriber_id to params
    params.append(subscriber_id)

    sql = f"""
        UPDATE subscribers
        SET {', '.join(update_fields)}
        WHERE id = %s
        RETURNING *
    """

    result = query_one(sql, tuple(params))
    if result:
        return dict(result)
    return None
